README
# Apy

Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/apy`. To experiment with that code, run `bin/console` for an interactive prompt.

TODO: Delete this and the text above, and describe your gem

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'apy'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install apy

## Usage

TODO: Write usage instructions here

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/apy.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG
## [Unreleased]

## [0.0.1] - 2021-05-01

- Initial release


test/apy_test.rb
# frozen_string_literal: true

require "test_helper"

class ApyTest < Minitest::Test
  include Apy::Calculable

  def test_calculable_weighted_harmonic_mean
    inputs = [
      [1000, 156.23],
      [1000, 156.30],
      [1000, 173.15],
      [1000, 188.72],
      [1000, 204.61],
      [1000, 178.23]
    ]
    actual = weighted_harmonic_mean(inputs)

    assert_in_delta 174.57, actual, 0.01
  end

  def test_calculable_compound
    monthly_actual = compound(1200, rate: 0.1, times: 12, terms: 1)
    weekly_actual = compound(1200, rate: 0.1, times: 52, terms: 1)
    daily_actual = compound(1200, rate: 0.1, times: 365, terms: 1)

    assert_in_delta 1325.66, monthly_actual, 0.01
    assert_in_delta 1326.07, weekly_actual, 0.01
    assert_in_delta 1326.19, daily_actual, 0.01
  end

  def test_calculable_dca_compound
    monthly_2y_actual = dca_compound([[1200, 0.1], [1200, 0.1]], times: 12)

    assert_in_delta 2790.125, monthly_2y_actual, 0.01
  end

  def test_interest_total
    interest = Apy::Interest.new(apy: 0.1)

    annual_actual = interest.total(1200)
    monthly_actual = interest.total(1200, times: 12)

    assert_in_delta 1320.0, annual_actual, 0.01
    assert_in_delta 1325.66, monthly_actual, 0.01
  end

  def test_interest_dca
    interest = Apy::Interest.new(apy: 0.1)

    annual_actual = interest.dca(1200)
    quarterly_actual = interest.dca(300, times: 4)
    monthly_actual = interest.dca(100, times: 12)

    assert_in_delta 1320.0, annual_actual, 0.01
    assert_in_delta 1277.64, quarterly_actual, 0.01
    assert_in_delta 1267.29, monthly_actual, 0.01
  end

  def test_loan_payment_size
    d1 = Date.parse "2020-01-01"
    d2 = Date.parse "2021-01-01"
    loan = Apy::Loan.new(1200, apy: 0.1)

    lump_sum_actual = loan.payment_size(start_date: d1, end_date: d2)
    recurring_actual = loan.payment_size(start_date: d1, end_date: d2, payments_per_term: 12)

    assert_in_delta 1320.0, lump_sum_actual
    assert_in_delta 110.0, recurring_actual
  end

  def test_loan_total_owed
    d1 = Date.parse "2020-01-01"
    d2 = Date.parse "2021-01-01"
    loan = Apy::Loan.new(1200, apy: 0.1)

    lump_sum_actual = loan.total_owed(start_date: d1, end_date: d2)

    assert_in_delta 1320.0, lump_sum_actual
  end

  def test_loan_amortized_payment_size
    d1 = Date.parse "2020-01-01"
    d2 = Date.parse "2021-01-01"
    loan = Apy::Loan.new(1200, apy: 0.1)

    lump_sum_actual = loan.amorized_payment_size(start_date: d1, end_date: d2)

    assert_in_delta 105.5, lump_sum_actual, 0.01
  end
end


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "apy"

require "minitest/autorun"


lib/apy/loan.rb
# frozen_string_literal: true

require_relative "interest"

module Apy
  # A loan object; Provides helpers for calculating various debt-related scenarios
  class Loan
    attr_reader :borrow, :apy

    # @param borrow [Numeric] Amount of money to be borrowed
    # @param apy [Float] The interest rate of the borrowed money
    def initialize(borrow, apy:)
      fail(ArgumentError, "apy must be a positive Float") unless apy.positive? && apy.is_a?(Float)

      @borrow = borrow
      @apy = apy
    end

    # Based on the borrow amount, calculate the payment size required to pay off the principal and accrued interest
    # @param start_date [Date] Date the loan will begin
    # @param end_date [Date] Date the loan will be fully paid off
    # @param times [Integer] The number of times per term interest is accrued; Defaults to 1 (flat rate)
    # @param days_per_term [Integer] The number of days per interest term
    # @param payments_per_term [Integer] The number of payments made towards the loan per term; Defaults to `times`
    # @example Payment size for 1200, interest accrued once, repaid in 1 payment across 365 days:
    #   d1 = Date.parse "2020-01-01"
    #   d2 = Date.parse "2021-01-01"
    #   Loan.new(1200, apy: 0.1).payment_size(start_date: d1, end_date: d2) == 1320.0
    # @example Payment size for 1200, interest accrued once, repaid in 12 payments across 365 days:
    #   Loan.new(1200, apy: 0.1).payment_size(start_date: d1, end_date: d2, payments_per_term: 12) == 110.0
    # @example Payment size for 1200, interest accrued 12 times, repaid in 12 payments across 365 days:
    #   Loan.new(1200, apy: 0.1).payment_size(start_date: d1, end_date: d2, times: 12) == 110.47
    def payment_size(start_date:, end_date:, times: 1, days_per_term: 365, payments_per_term: times)
      total_owed(
        start_date: start_date,
        end_date: end_date,
        times: times,
        days_per_term: days_per_term
      ) / (payments_per_term * Interest.get_term_size(start_date, end_date, days_per_term))
    end

    # Get the total amount owed, based on the start & end date
    # @param start_date [Date] Date the loan will begin
    # @param end_date [Date] Date the loan will be fully paid off
    # @param times [Integer] The number of times per term interest is accrued; Defaults to 1 (flat rate)
    # @param days_per_term [Integer] The number of days per interest term
    # @note Because the constructor accepts an APY for the year, an adjusted rate is used here based on days_per_term
    # @see Interest#total
    # @see #adjusted_apy
    def total_owed(start_date:, end_date:, times: 1, days_per_term: 365)
      Apy::Interest.new(
        apy: adjusted_apy(days_per_term),
        start_date: start_date,
        end_date: end_date,
        days_per_term: days_per_term
      ).total(borrow, times: times)
    end

    # Similar to payment_size, except interest accrues based on the remaining debt
    # @todo Finish this
    # @todo Once finished, make #payment_size accept less args (lump-sum, _only_ accept payment count)
    def amortized_payment_size(start_date:, end_date:, times: 1, days_per_term: 365, payments_per_term: times)
      fail
    end

    # @todo Finish this
    # @todo Once finished, make #total_owed accept less args (lump-sum, _only_ accept payment count)
    def amortized_total_owed(start_date:, end_date:, times: 1, days_per_term: 365, payments_per_term: times)
      fail
    end

    private

    # `apy / (365 / days_per_term)`
    def adjusted_apy(days_per_term)
      apy / (365 / days_per_term)
    end
  end
end


lib/apy/interest.rb
# frozen_string_literal: true

require "date"
require_relative "calculable"

module Apy
  # An interest object; Given an apy & date range, prepares various methods to calculate with
  class Interest
    include Apy::Calculable

    attr_reader :apy

    # @param apy [Float] Annual/Average Percent Yield -- an expected pct return over the course of a _year_
    # @param start_date [Date] The beginning of the interest period; Defaults to today
    # @param end_date [Date] The end of the interest period; Defaults to 1y from today
    # @param days_per_term [Integer] Days per compound interest term; Defaults to 365 days
    #
    # @example A 10% APY, intended to be used in calculations over 2y:
    #   d1 = Date.parse "2020-01-01"
    #   d2 = Date.parse "2022-01-01"
    #   Interest.new(apy: 0.1, start_date: d1, end_date: d2)
    # @example An interest object _yielding 10% over the course of 2y_:
    #   d1 = Date.parse "2020-01-01"
    #   d2 = Date.parse "2022-01-01"
    #   Interest.new(apy: 0.1, start_date: d1, end_date: d2, days_per_term: 730)
    # @example The above example, but instantiated with an apy of 5%:
    #   d1 = Date.parse "2020-01-01"
    #   d2 = Date.parse "2022-01-01"
    #   Interest.new(apy: 0.05, start_date: d1, end_date: d2)
    def initialize(apy:, start_date: Date.today, end_date: Date.today.next_year, days_per_term: 365)
      fail(ArgumentError, "apy must be a positive Float") unless apy.positive? && apy.is_a?(Float)

      @apy = apy
      @terms = Interest.get_term_size(start_date, end_date, days_per_term)
    end

    class << self
      # @param days_per_term [Integer] Number of days that pass before an entire term ends. Defaults to 365
      # @return [Integer] The actual number of terms completed
      def get_term_size(start_date, end_date, days_per_term)
        ((end_date - start_date) / days_per_term).round
      end
    end

    # Given a principal amount, return the ending balance
    # @param principal [Numeric] Initial investment
    # @param times [Integer] The number of times per term interest is accrued; Defaults to 1 (flat rate)
    # @see Calculable#compound
    def total(principal, times: 1)
      compound(principal, rate: apy, times: times, terms: @terms)
    end

    # Given a series of investments, calculate the DCA return
    # @param in_per_split [Numeric] Value of newly invested funds per term; Will be zipped with #apy & term size
    # @param times [Integer] The number of times per term interest is accrued; Defaults to 1 (flat rate)
    # @note This assumes you wish to maximize dca returns; the in_per_split is deposited before the interest is calculated
    # @note If this method is too inflexible for your use case, you should use the {Calculable} module directly
    # @see Calculable#dca_compound
    # @example Investing 3.29 everyday for a year:
    #   Apy::Interest.new(apy: 0.1).dca(3.29, times: 365) == 1263.12
    # @example Investing 23.08 every week for a year:
    #   Apy::Interest.new(apy: 0.1).dca(23.08, times: 52) == 1263.43
    # @example Investing 100 every month for a year:
    #   Apy::Interest.new(apy: 0.1).dca(23.08, times: 12) == 1267.29
    # @example Investing 300 every quarter for a year:
    #   Apy::Interest.new(apy: 0.1).dca(300, times: 4) == 1277.64
    # @example Investing 600 semi-annualy for a year:
    #   Apy::Interest.new(apy: 0.1).dca(600, times: 2) == 1292.66
    def dca(in_per_split, times: 1)
      split = @terms * times
      adjusted_apy = apy / times
      range = split.times.map { [in_per_split, adjusted_apy] }

      dca_compound(range, times: times)
    end
  end
end


lib/apy/calculate.rb
# frozen_string_literal: true

module Apy
  module Calculate
    # Calculate the weighted harmonic mean for a set of values
    # (∑w) / (∑w/i)
    # @param matrix [Array<Array<(Numeric, Numeric)>>] Matrix whose size is the number of investments, with the following array elements: idx0 invested amount, idx1 purchase price of the asset
    # @todo Can make this a bit more performant with `inject`
    # @example Providing a set of DCA values
    #   dca = [[1000,156.23],[1000,156.3], [1000,173.15],[1000,188.72], [1000,204.61],[1000,178.23]]
    #   weighted_harmonic_mean(dca)
    #   # => 174.5655590168175
    def weighted_harmonic_mean(matrix)
      sum = matrix.sum { |(n, _)| n }.to_f

      total_weight = []
      weighted_values = []

      matrix.each do |(investment, price)|
        weight = investment / sum

        total_weight << weight
        weighted_values << (weight / price)
      end

      total_weight.sum / weighted_values.sum
    end

    # Simple compound, assuming no additional investment over successive maturity terms
    # @param principal [Numeric] Initial investment
    # @param rate [Float] Expected interest rate for the length of the period
    # @param times [Integer] Times the interest will be paid out per term
    # @param terms [Integer] Number of terms
    #
    # @example Given a "10% APY", with interest paid monthly (1y maturity date):
    #   compound(1200, rate: 0.1, times: 12, terms: 1) == 1325.66
    #
    # @example Given a "0.1923% WPY", with interest paid weekly (1w maturity date):
    #   compound(1200, rate: 0.1, times: 52, terms: 1) == 1326.07
    #   compound(1200, rate: 0.001923, times: 1, terms: 52) == 1326.07
    #
    # @example Given a "0.0274% DPY", with interest paid daily (1d maturity date):
    #   compound(1200, rate: 0.1, times: 365, terms: 1) == 1326.19
    #   compound(1200, rate: 0.000274, times: 1, terms: 365) == 1326.20
    def compound(principal, rate:, times:, terms:)
      total_rate = 1 + (rate / times)

      principal * (total_rate**(times * terms))
    end

    # "DCA" compound, assuming a recurring investment continuously added to the principal amount, this new amount _additionally_ compounded for the next period
    # @param matrix [Array<Array(Numeric, Numeric)>] Matrix whose size is the number of terms; Each array item has the following elements: idx0 additional investment, idx1 the expected rate for the term
    # @param times [Integer] Times the interest will be paid out per term
    # @example Continuously investing 1200 dollars a year into a contract with a "10% APY", interest paid out once a month
    #   dca_compound([[1200, 0.1], [1200, 0.1]], times: 12) == 2790.125
    #
    # @todo Clean this up, there is most likely an optimized formula for this 🤨
    # @see #compound
    def dca_compound(matrix, times:)
      result = matrix.each_with_object(
        total: 0,
        interest: 0,
        data: {
          0 => {
            ytd: 0,
            in: 0,
            interest: 0
          }
        }
      ).with_index do |(ary, out), i|
        additional_investment, rate = ary
        prev_ytd = out[:data][i][:ytd]

        to_compound = prev_ytd + additional_investment

        current = compound(to_compound, rate: rate, times: times, terms: 1)
        interest_this_period = current - to_compound

        out[:total] += additional_investment + interest_this_period
        out[:interest] += interest_this_period
        out[:data][i + 1] = {ytd: current, in: additional_investment, interest: interest_this_period}
      end

      result.fetch(:total)
    end
  end
end


lib/apy/calculable.rb
# frozen_string_literal: true

module Apy
  module Calculable
    # Calculate the weighted harmonic mean for a set of values
    # (∑w) / (∑w/i)
    # @param matrix [Array<Array<(Numeric, Numeric)>>] Matrix whose size is the number of investments, with the following array elements: idx0 invested amount, idx1 purchase price of the asset
    # @todo Can make this a bit more performant with `inject`
    # @example Providing a set of DCA values
    #   dca = [[1000,156.23],[1000,156.3], [1000,173.15],[1000,188.72], [1000,204.61],[1000,178.23]]
    #   weighted_harmonic_mean(dca)
    #   # => 174.5655590168175
    def weighted_harmonic_mean(matrix)
      sum = matrix.sum { |(n, _)| n }.to_f

      total_weight = []
      weighted_values = []

      matrix.each do |(investment, price)|
        weight = investment / sum

        total_weight << weight
        weighted_values << (weight / price)
      end

      total_weight.sum / weighted_values.sum
    end

    # Simple compound, assuming no additional investment over successive maturity terms
    # @param principal [Numeric] Initial investment
    # @param rate [Float] Expected interest rate for the length of the period
    # @param times [Integer] Times the interest will be paid out per term
    # @param terms [Integer] Number of terms
    #
    # @example Given a "10% APY", with interest paid monthly (1y maturity date):
    #   compound(1200, rate: 0.1, times: 12, terms: 1) == 1325.66
    #
    # @example Given a "0.1923% WPY", with interest paid weekly (1w maturity date):
    #   compound(1200, rate: 0.1, times: 52, terms: 1) == 1326.07
    #   compound(1200, rate: 0.001923, times: 1, terms: 52) == 1326.07
    #
    # @example Given a "0.0274% DPY", with interest paid daily (1d maturity date):
    #   compound(1200, rate: 0.1, times: 365, terms: 1) == 1326.19
    #   compound(1200, rate: 0.000274, times: 1, terms: 365) == 1326.20
    def compound(principal, rate:, times:, terms:)
      total_rate = 1 + (rate / times)

      principal * (total_rate**(times * terms))
    end

    # "DCA" compound, assuming a recurring investment continuously added to the principal amount, this new amount _additionally_ compounded for the next period
    # @param matrix [Array<Array(Numeric, Numeric)>] Matrix whose size is the number of terms; Each array item has the following elements: idx0 additional investment, idx1 the expected rate for the term
    # @param times [Integer] Times the interest will be paid out per term
    # @example Continuously investing 1200 dollars a year into a contract with a "10% APY", interest paid out once a month
    #   dca_compound([[1200, 0.1], [1200, 0.1]], times: 12) == 2790.125
    #
    # @todo Clean this up, there is most likely an optimized formula for this 🤨
    # @see #compound
    def dca_compound(matrix, times:)
      result = matrix.each_with_object(
        total: 0,
        interest: 0,
        data: {
          0 => {
            ytd: 0,
            in: 0,
            interest: 0
          }
        }
      ).with_index do |(ary, out), i|
        additional_investment, rate = ary
        prev_ytd = out[:data][i][:ytd]

        to_compound = prev_ytd + additional_investment

        current = compound(to_compound, rate: rate, times: times, terms: 1)
        interest_this_period = current - to_compound

        out[:total] += additional_investment + interest_this_period
        out[:interest] += interest_this_period
        out[:data][i + 1] = {ytd: current, in: additional_investment, interest: interest_this_period}
      end

      result.fetch(:total)
    end
  end
end
